File: Two_Pointers/986.Interval-List-Intersections/986.Interval-List-Intersections.java
class Solution {
    public int[][] intervalIntersection(int[][] firstList, int[][] secondList) {
        List<int[]> ret = new ArrayList<>(); 
        int i = 0;
        int j = 0;
        while (i < firstList.length && j < secondList.length) {
            int start = Math.max(firstList[i][0], secondList[j][0]);
            int end = Math.min(firstList[i][1], secondList[j][1]);
            if (end >= start) {
                ret.add(new int[]{start, end});
            }
            if (firstList[i][1] < secondList[j][1]) {
                i++;
            } else {
                j++;
            }
        }
        return ret.toArray(new int[ret.size()][]);
    }
}

File: Two_Pointers/986.Interval-List-Intersections/986.Interval-List-Intersections.ts
function intervalIntersection(
  firstList: number[][],
  secondList: number[][]
): number[][] {
  if (firstList.length === 0 || secondList.length === 0) {
    return [];
  }
  let ans: number[][] = [];
  let i = 0;
  let j = 0;
  while (i < firstList.length && j < secondList.length) {
    let left = Math.max(firstList[i][0], secondList[j][0]);
    let right = Math.min(firstList[i][1], secondList[j][1]);
    if (left <= right) {
      ans.push([left, right]);
    }
    // 哪个右区间元素较小，就向前移动一位
    if (firstList[i][1] < secondList[j][1]) {
      i++;
    } else {
      j++;
    }
  }
  return ans;
}
